lisa/sut_orchestrator/baremetal/platform_.py

```diff
@@ -0,0 +1,138 @@
+class SourceBuildSchema:
+    source: Optional[SourceSchema] = None
+
+
+@dataclass_json()
+@dataclass
+class BareMetalPlatformSchema:
+    source: Optional[SourceSchema] = field(default=None)
+    cluster: ClusterSchema = field(default_factory=ClusterSchema)
+
+
+class BareMetalPlatform(Platform):
+    def __init__(
+        self,
+        runbook: schema.Platform,
+    ) -> None:
+        super().__init__(runbook=runbook)
+
+    @classmethod
+    def type_name(cls) -> str:
+        return BAREMETAL
+
+    @classmethod
+    def supported_features(cls) -> List[Type[feature.Feature]]:
+        return [features.StartStop]
+
+    def _initialize(self, *args: Any, **kwargs: Any) -> None:
+        baremetal_runbook: BareMetalPlatformSchema = self.runbook.get_extended_runbook(
+            BareMetalPlatformSchema
+        )
+        assert baremetal_runbook, "platform runbook cannot be empty"
+        self._baremetal_runbook = baremetal_runbook
+
+    def _prepare_environment(self, environment: Environment, log: Logger) -> bool:
+        return True
+
+    def _deploy_environment(self, environment: Environment, log: Logger) -> None:
+        # copy build (shared, check if it's copied)
+        source_path: str = ""
+        if self._baremetal_runbook.source and self._baremetal_runbook.source:
+            source_factory = Factory[Source](Source)
+            source = source_factory.create_by_runbook(self._baremetal_runbook.source)
+            self._log.debug(f"found build '{source.type_name()}', to expand runbook.")
+            source.initialize()
+            source_path = source.source_path
+
+        ready_checker_factory = Factory[ReadyChecker](ReadyChecker)
+        cluster_factory = Factory[Cluster](Cluster)
+        ip_getter_factory = Factory[IpGetterChecker](IpGetterChecker)
+        ip_address: str
+        cluster = self._baremetal_runbook.cluster
+        # ready checker cleanup
+        if cluster.ready_checker:
+            ready_checker_runbook = ready_checker_factory.create_by_runbook(
+                cluster.ready_checker
+            )
+            ready_checker_runbook.clean_up()
+
+        # copy build if source exists
+        cluster_runbook = cluster_factory.create_by_runbook(cluster)
+        cluster_runbook.initialize(source_path=source_path)
+
+        # deploy cluster
+        cluster_runbook.deploy()
+
+        # ready checker
+        if cluster.ready_checker:
+            ready_checker_runbook = ready_checker_factory.create_by_runbook(
+                cluster.ready_checker
+            )
+            check_till_timeout(
+                lambda: ready_checker_runbook.is_ready() is True,
+                timeout_message="wait for ready check ready",
+            )
+
+        # get ip address
+        if cluster.ip_getter:
+            ip_getter_runbook = ip_getter_factory.create_by_runbook(cluster.ip_getter)
+            ip_address = ip_getter_runbook.get_ip()
+
+        if cluster_runbook.runbook.client.ip_address:
```

Comment:
If the ip_address is set, it should be used before trying to use ip_getter. If none of them set, there should be an exception. So it's like,

```python
if cluster_runbook.runbook.client.ip_address:
    ...
elif cluster.ip_getter:
    ...
else:
    raise LISAException(...)
```